warehouse/macaroons/security_policy.py

```diff
@@ -67,16 +68,18 @@ def _extract_http_macaroon(request):
     return None
 
 
-@implementer(IAuthenticationPolicy)
-class MacaroonAuthenticationPolicy(CallbackAuthenticationPolicy):
-    def __init__(self, callback=None):
-        self.callback = callback
-
-    def unauthenticated_userid(self, request):
+@implementer(ISecurityPolicy)
+class MacaroonSecurityPolicy:
+    def identity(self, request):
         # If we're calling into this API on a request, then we want to register
         # a callback which will ensure that the response varies based on the
         # Authorization header.
         request.add_response_callback(add_vary_callback("Authorization"))
+        request.authentication_method = AuthenticationMethod.MACAROON
+
+        # Macaroon authentication can only be used for uploading
+        if request.matched_route.name not in ["forklift.legacy.file_upload"]:
+            return False
```

Comment:
Flagging this for review as well: we didn't actually have this before, which makes me think that our authn/z for macaroons might have been a little too permissive (we were probably allowing anything with `context=Project` that matched the token's actual permissions).

Reply:
Yes, this is known: https://github.com/pypa/warehouse/issues/7266

Reply:
(I'm also not convinced that we need this, at least in this PR).

Reply:
Totally missed that check! I can remove this route check then, since it's a little out of place in the AuthN section.

Reply:
[0bc2083](https://github.com/pypa/warehouse/pull/11218/commits/0bc208360ef2f1f94146e35396e4ef981a4b77c1)